src/app_service/service/message/service.nim

```diff
@@ -176,26 +176,34 @@ QtObject:
         self.events.emit(SIGNAL_MESSAGE_EDITED, data)
 
     for i in 0 ..< chats.len:
+      let chatId = chats[i].id
+
       if(chats[i].chatType == ChatType.Unknown):
-        error "error: new message with an unknown chat type received", chatId=chats[i].id
+        error "error: new message with an unknown chat type received", chatId=chatId
         continue
 
       # Ignore 1-1 chats for which we are not contact
-      if (chats[i].chatType == ChatType.OneToOne and not self.contactService.getContactById(chats[i].id).isContact):
+      if (chats[i].chatType == ChatType.OneToOne and not self.contactService.getContactById(chatId).isContact):
         continue
 
-
       var chatMessages: seq[MessageDto]
       for msg in messages:
-        if (msg.localChatId == chats[i].id):
-          chatMessages.add(msg)
+        if (msg.localChatId != chatId):
+          continue
+
+        # Ignore messages older than current chat cursor
+        if (self.msgCursor.hasKey(chatId)):
+          let currentChatCursor = self.msgCursor[chatId]
+          let msgCursorValue = initCursorValue(msg.id, msg.clock)
+          if (not currentChatCursor.isLessThan(msgCursorValue)):
+            currentChatCursor.makeObsolete()
```

Comment:
For my understanding, making the cursor obsolete means we'll still be checking the last cursor on line 196, but when we scroll, it starts again? Can you explain so I can understand? Thanks

Reply:
It is exactly as you understand. Making the cursor obsolete means we will try to load messages the next time we scroll, even though we hit the top of the database before.

Time A (we loaded all messages):
|msg|in-database|in-app|cursor|cursor_most_recent|
|-|-|-|-|-|
4|yes|yes|X|yes (we hit the top of the database) |
5|yes|yes|||
6|yes|yes|||

We try to scroll but we don't load anything, because we know the cursor is the most recent.

Time B (we fetch some messages):
|msg|in-database|in-app|cursor|cursor_most_recent|
|-|-|-|-|-|
2|yes|no|||
3|yes|no|||
4|yes|yes|X|yes|
5|yes|yes|||
6|yes|yes|||

In this case scroll would not work, even though we have new messages in database.

We need to make the cursor obsolete (not most recent) aka `currentChatCursor.makeObsolete()`:

|msg|in-database|in-app|cursor|cursor_most_recent|
|-|-|-|-|-|
2|yes|no|||
3|yes|no|||
4|yes|yes|X|no|
5|yes|yes|||
6|yes|yes|||

in this case the scroll will work properly:

|msg|in-database|in-app|cursor|cursor_most_recent|
|-|-|-|-|-|
2|yes|yes|X|yes|
3|yes|yes|||
4|yes|yes|||
5|yes|yes|||
6|yes|yes|||

Reply:
That makes it very clear, thanks for the explanation!